src/hooks/transactions/useReceiver.ts

```diff
@@ -14,43 +14,110 @@
-
-  const validateAccount = (receiver: string) => {
-    try {
-      const { address, formatFound } = getReceiverAddress({
-        getChainBySS58Prefix,
-        targetChainDetails,
-        sourceChainDetails,
-        receiverAddress: receiver
-      });
-
-      return { formatFound, formattedAccount: address };
-    } catch (e) {
-      logger.error(e.message);
-      if (e.message === INCORRECT_FORMAT) {
-        return { formatFound: e.message, formattedAccount: receiver };
+  const setReceiver = useCallback(
+    (address: string | null) => dispatchTransaction(TransactionActionCreators.setReceiverAddress(address)),
+    [dispatchTransaction]
+  );
+
+  const setUnformattedReceiver = useCallback(
+    (address: string | null) => dispatchTransaction(TransactionActionCreators.setUnformattedReceiverAddress(address)),
+    [dispatchTransaction]
+  );
+
+  const validateAccount = useCallback(
+    (receiver: string) => {
+      try {
+        const { address, formatFound } = getReceiverAddress({
+          getChainBySS58Prefix,
+          targetChainDetails,
+          sourceChainDetails,
+          receiverAddress: receiver
+        });
+
+        return { formatFound, formattedAccount: address };
+      } catch (e) {
+        logger.error(e.message);
+        if (e.message === INCORRECT_FORMAT) {
+          return { formatFound: e.message, formattedAccount: receiver };
+        }
       }
+    },
+    [getChainBySS58Prefix, sourceChainDetails, targetChainDetails]
+  );
+
+  const reset = useCallback(() => {
+    dispatchTransaction(TransactionActionCreators.reset());
+    setShowBalance(false);
+  }, [dispatchTransaction]);
+
+  const onReceiverChange = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      const receiver = event.target.value;
+      reset();
+      setUnformattedReceiver(receiver);
+      if (!receiver) {
+        return;
+      }
+
+      const { formattedAccount, formatFound } = validateAccount(receiver)!;
+      setFormatFound(formatFound);
+      if (formatFound === INCORRECT_FORMAT) {
+        dispatchTransaction(TransactionActionCreators.setValidationError('Invalid address'));
+        return;
+      }
+
+      if (formatFound === GENERIC) {
+        dispatchTransaction(TransactionActionCreators.setGenericAccount(receiver));
+        return;
+      }
+
+      if (formatFound === targetChain) {
+        setReceiver(formattedAccount);
+        setShowBalance(true);
+        return;
+      }
+
+      if (formatFound === sourceChain) {
+        dispatchTransaction(TransactionActionCreators.setDerivedAccount(formattedAccount));
+        setReceiver(receiver);
+        return;
+      }
+
+      dispatchTransaction(
+        TransactionActionCreators.setValidationError(`Unsupported address SS58 prefix: ${formatFound}`)
+      );
+    },
+    [dispatchTransaction, reset, setReceiver, setUnformattedReceiver, sourceChain, targetChain, validateAccount]
+  );
```

Comment:
It's a bit hard to tell what the specific goal of the PR currently is cause it's still a draft, so I believe there is more work expected to be done here, but IMHO this entire function should simply be part of the reducer.

I don't see a reason why we send multiple actions (`reset` and then something else) in pretty much one function frame - this to me sounds like a reducer anti-pattern (mentioned that in some other PR - the actions should be as course-grained and domain-specific as possible).

`onReceiverChange` callback should literally just send one action `setUnformattedReceiver` and the rest (validation, format detection logic, etc, should be part of the reducer). I guarantee that this will make the code much easier to understand.

As mentioned earlier, using reducers without placing domain logic there (and instead keeping it in components/hooks) is misusing reducers and just adds unnecessary boilerplate of dispatching transactions - instead top-level `useState` with values&setters passed through context would be much simpler.

Overall if you are not intending to move logic to reducers at all, then let's keep the code as-is, cause having some imho half-baked solutions (where some stuff happens in the reducer and some logic is in hooks) is going to complicate the code instead of improving it.

Reply:
I think this comment is outdated now since the approach of the PR changed drastically.